### DISCUSSION TITLE
post/overcome-spring-aop-with-kotlin/

### DISCUSSION THREAD

**Discussion by giscus** [kakaopay/techlog-comments#36] [Announcements]
# post/overcome-spring-aop-with-kotlin/

Kotlin의 문법적 기능을 사용해서 Spring AOP 아쉬운 점을 극복한 경험을 공유합니다.

https://tech.kakaopay.com/post/overcome-spring-aop-with-kotlin/

**Comment by 134130** [comment]
코틀린 및 Spring을 입문하며 `companion object` 혹은 `private fun`에 `@Transactional`을 붙이고 싶은 경우가 있었는데, 해당 글이 많은 도움이 됐습니다. 감사합니다.

**Comment by vcnc-jara** [comment]
@transactional 에서 isolation level 같은 걸 넣을 수 있는 걸로 아는데, 그런 파라미터는 어떻게 만들면 사용할 수 있을까요?

**Reply by cglee079** [reply]
안녕하세요, 답변이 너무 늦어 죄송합니다!

어노테이션 변수는 런타임 단계에서 수정할 수 없어, 파라미터로 만들고 사용할 수 없어서요
위의 작성처럼`Tx.repetableRead{}` 와 같은 방식으로 별도의 함수를 정의해서 사용하고 있습니다.

감사합니다.

**Comment by sosow0212** [comment]
학습하는 데 도움이 많이 됐습니다. 좋은 글 감사합니다!

**Comment by J-mung** [comment]
@Cachable에서 key 인자를 SpEL으로 사용하면서 발생할 수 있는 불일치 문제를 동일하게 느끼고 있었습니다. 
뜻하지 않게 해당 글로 해결법을 얻는 등, 많은 도움을 얻게 되어 읽는 내내 즐거웠습니다.
감사합니다 :D

**Comment by Vorschlag-bit** [comment]
너무 좋은 방식이지만 정적 컴포넌트라서 단위 테스트 시, mocking 및 stubbing하기가 굉장히 애매해지던데 혹시 이런 문제는 어떻게 해결하셨는지 알 수 있을까요..?

**Reply by cglee079** [reply]
안녕하세요! 

Trailing Lamdas 로 구현된 AOP 코드 자체를 단위 테스트로 작성하시는 것은 어려움이 없으실 것이라 생각합니다.

아마 말씀하신 의도는, 해당 AOP를 사용하는 코드의 단위 테스트를 말씀하시는것으로 이해가 되는데요..!

글 내용의 Tx를 예시로든다며 TxAdvice를 목킹하여,  Tx에 주입하는 테스트 초기화 작업을 하고 있습니다.
물론 이를 위해서 별도의 Tx에 TxAdvice 주입구를 열어주고 사용하고 있습니다.

감사합니다.

**Comment by earlgrey02** [comment]
혹시 `Tx` 클래스를 `@Transactional`이 아닌 `PlatformTransactionManager` 기반으로 구현하는 방법은 어떻게 생각하시나요??
지금은 어노테이션 변수(컴파일 상수)의 한계로 트랜잭션 설정(`readOnly`나 `propagation` 등)에 따라 각각 함수를 따로 구현하시는 걸로 알고 있습니다.
만약 `PlatformTransactionManager`로 programmatic 하게 구현하면 하나의 함수만으로 인자를 다르게 해서 런타임에도 트랜잭션 설정을 유연하게 제어할 수 있을 것 같은데, 이에 대한 의견을 여쭤보고 싶습니다..!

**Reply by 134130** [reply]
제 기억에는, 테스트쪽에서 TestTransaction 기능을 제공할 때, 어노테이션을 기반으로만 처리해줬던 것 같아요.

함수가 많아지는 부분은, 어차피 외부로 노출되지 않는 영역이 많아지는 것이고, 외부에서는 단일 인터페이스로 유지하면 되는 영역이라 저는 크게 중요하지 않다고 생각합니다.

**Reply by earlgrey02** [reply]
실제로 `TransactionalTestExecutionListener`가 `AnnotationTransactionAttributeSource` 기반으로 동작하다 보니, 말씀해주신 것처럼 programmatic 하게 구현된 `Tx`는 `TestTransaction`과 함께 사용하기 어렵다는 한계가 있겠네요.. 대신 답변 감사드립니다!

**Reply by earlgrey02** [reply]
근데 확인해보니까 `TestTransaction`이랑은 관련이 없을 것 같습니다. `TestTransaction`이 어노테이션 기반이라는 것은 테스트 메서드의 `@Transactional` 등록 여부로 `TestTransaction`의 생성 여부를 결정한다는 부분이고, `TestTransaction` 내에서는 어떤 방식으로 구현된 `Tx`든 정상적으로 동작(롤백)하는 것 같더라고요.

그래서 저는 그냥 두 방식(선언적인 방식과 programmatic 방식)은 단순 스타일 차이라고 결론 내렸습니다..ㅎ

**Reply by 134130** [reply]
앗 그렇군요 ㅎㅎ

오래전에 뭔가 테스트코드쪽에서 문제가 되었던 기억이 있는데, 저희가 Transaction 어노테이션 기반으로 추가로 적용한 AOP 일 수 도 있습니다 ㅎㅎ

공유해주셔서 감사합니다~
